Enable GEOPM service to automatically restart effectively

cmcantalupo opened the pull request:
- Fixes #2077
- Fixes #2078 
- Fixes #2085

Unit tests:
- [x] test_default_creation
- [x] test_creation_bad_perms
- [x] test_creation_bad_owner
- [x] test_creation_sessions
- [x] test_creation_bad_session_perms
- [x] test_creation_bad_session_owner
- [x] test_update_reload f8948fcae
- [x] test_add_remove_client ad40d2a
- [x] test_write_client c582750
- [x] test_batch_server 823bb34
- [x] test_watch_id b5e407b

cmcantalupo committed:
Enable geopm-service Restart systemd feature

- Add init file for geopmdpy_test directory
- Parse session files in /var/run at geopmd startup
- Valides session file properties
- Create new geopmdpy module called varrun that
  includes the ActiveSession class
- Add Restart=always to the systemd unit config
- Add jsonschema python requirements to spec file
  and python package
- Reset watch_id for clients pids that persist
  through a restart
- Update documentation based on feedback
- Fix bug in time comparison for recycled PID case
- Add several TODO comments where there are disparities
  between the implementation and the documentation.
- Change ActiveSessions.get_mode to ActiveSessions.is_write_client
- Add first two units test for ActiveSessions
- Add rest of doc strings for ActiveSessions public members
- Add stubs for required unit tests
- Add comments describing each unit test for ActiveSessions
- Add test_watch_id fixture and fix other tests so that they execute

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz committed:
Refactor service.py tests to use ActiveSessions mock

- Introduce new mocking pattern for objects that maintain
  state.  Monkey patching won't work here.
- Resolves #2086.
- Remove bad test constructor pattern.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo committed:
Add new unit test fixtures to TestActiveSessions

- Add test_batch_server test
- Add test_write_client

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz committed:
Populate TestActiveSessions

- test_update_reload
- test_add_remove_client
- Refactor check_getters into a class method

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

Krebrov001 committed:
Correct ActiveSessions implementation to reflect new tests

Refactor the `__init__()` of `ActiveSessions` in varrun.py

Make the conditions which determine how and when to create a directory
to hold the session files, more robust.

The directory is created if it doesn't already exist, or if it exists in
an invalid state, which could be one of the following:
 - It's a link not a directory.
 - It's any other kind of file not a directory.
 - The directory's permissions is not what we wanted.
 - The directory's user owner is not what we wanted.
 - The directory's group owner is not what we wanted.

In that case it is renamed, and a new directory is created in it's place
with the desired name.

This commit also contains modifications to `_load_session_file()` to
facilitate printing extra verbose warning messages.

Added new tests to TestActiveSessions.py

 - test_creation_link_not_dir()
 - test_creation_file_not_dir()
 - test_creation_bad_perms()
 - test_creation_bad_user_owner()
 - test_creation_bad_group_owner()
 - test_creation_sessions()
 - test_creation_bad_session_perms()
 - test_creation_bad_session_user_owner()
 - test_creation_bad_session_group_owner()

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

cmcantalupo committed:
Rename invalid session files, do not unlink, or ignore

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

Krebrov001 committed:
Add tests checking how varrun.py deals with arbitrary text and json files.

Checks added to the `test_creation_json()`:

 - json_good_example, Valid
 - json_empty_signals_controls, Valid
 - json_no_mode, Invalid
 - json_bad_mode, Invalid
 - json_wrong_data_types, Invalid
 - json_additional_properties, Invalid
 - json_list_of_2, Invalid
 - json_empty_dictionary, Invalid
 - string_empty_file, Invalid
 - string_typos_json, Invalid
 - string_c_code, Invalid

Modify `create_json_file()` to take the contents as another parameter,
which can be a json dictionary or a string.

Remove test_creation_sessions

Add test_default_creation and test_creation_json to the Makefile.mk

Also add a fix enabling varrun.py to robustly process files which do not
contain valid json (previously there were no checks on processing files,
only for checking the schema).

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
I added Brad's new tests to the Makefile.mk

 - test_update_reload
 - test_add_remove_client
 - test_write_client
 - test_batch_server
 - test_watch_id

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

cmcantalupo commented on service/geopmdpy/varrun.py:
```diff
@@ -0,0 +1,519 @@
+        to a directory which is not accessible, the directory will be
+        renamed and a warning is printed to the syslog. The warning
+        message will display the reason why the directory was not
+        accessible, display the permissions or user and group of the directory being
```
ws

bgeltz commented on service/setup.py:
```diff
@@ -76,7 +76,8 @@
                     'psutil>=5.6.2',
-                    'dasbus>=1.5.0']
+                    'dasbus>=1.5.0',
+                    'jsonschema>=2.6.0']
```
We need to start messaging to Slack when the Python requirements change.  This ends up being a minor headache for most folks to deal with, and I don't want anyone to be surprised.

bgeltz commented on service/geopm.service:
```diff
@@ -37,6 +37,9 @@ Environment=PYTHONUNBUFFERED=true
 ExecStart=/usr/bin/geopmd
+Restart=always
+StartLimitInterval=0
+StartLimitBurst=5
```
AFAICT, the `StartLimit*` parameters belong in the `[Unit]` section, not the `[Service]` section.

https://www.freedesktop.org/software/systemd/man/systemd.service.html#Restart=
https://www.freedesktop.org/software/systemd/man/systemd.unit.html#StartLimitIntervalSec=interval

Additionally, the first option is `StartLimitIntervalSec` not `StartLimitInterval` according to the above docu.

bgeltz commented on service/geopmdpy/varrun.py:
```diff
@@ -0,0 +1,519 @@
+        renamed and a warning is printed to the syslog. The warning
+        message will display the reason why the directory was not
+        accessible, display the permissions or user and group of the directory being
+        deleteted.
```
```suggestion
        deleted.
```

bgeltz commented on service/geopmdpy/varrun.py:
```diff
@@ -0,0 +1,519 @@
+        accessible, display the permissions or user and group of the directory being
+        deleteted.
+
+        When an ActiveSessions object is created and the directory
+        already exists the owner of the directory will be set to the
+        geopmd user and the permissions will be set to mode 0o700.
+        All files matching the pattern
```
Is this still accurate, or are we deleting and re-creating?

cmcantalupo replied:
We are moving any files that do not meet requirements to uuid4-tmp versions

Krebrov001 commented on service/geopmdpy/varrun.py:
```diff
@@ -0,0 +1,519 @@
+        self._sessions[client_pid].pop('batch_server')
+        self._update_session_file(client_pid)
+
+    def _get_session_path(self, client_pid):
```
Docstring should be here.

Krebrov001 commented on service/geopmdpy/varrun.py:
```diff
@@ -0,0 +1,519 @@
+    def _get_session_path(self, client_pid):
+        return f'{self._VAR_PATH}/session-{client_pid}.json'
+
+    def _update_session_file(self, client_pid):
```
Docstring should be here.

Krebrov001 commented on service/geopmdpy/varrun.py:
```diff
@@ -0,0 +1,519 @@
+            json.dump(sess, fid)
+        os.rename(session_path_tmp, session_path)
+
+    def _load_session_file(self, sess_path):
```
Docstring should be here.

bgeltz committed:
Tweak JSON schema, simplify tests, and address feedback

- Require mode of session JSON to be one of 'r' or 'rw'.
- Add missing docstrings.
- Fixup mock_err assertions.
- Correct 'bad' data used in test.
- Align code and documentation.
- Update session file at the end of load.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo commented on service/integration/test/test_su_restart.sh:
```diff
@@ -0,0 +1,107 @@
+    user to read and write to bits 8-15 of the MSR_PERF_CTL register
+    which controls the maximum frequency of the core.  The test saves
+    the existing settings so they can be restored at the end of the
+    test.  Access to the control is removed from the list, and it is
```
This needs to mention the restart!

bgeltz commented on service/geopmdpy_test/TestActiveSessions.py:
```diff
@@ -0,0 +1,648 @@
+                mock.call(f'Warning: <geopm-service> {sess_path} has wrong permissions, it will be renamed to {renamed_path}'),
+                mock.call(f'Warning: <geopm-service> the wrong permissions were {oct(bad_user.st_mode)}')
+            ]
+            for cc, mc in zip(calls, mock_err.call_args_list):
+                self.assertEqual(cc, mc)
```
```suggestion
            mock_err.assert_has_calls(calls)
```

bgeltz commented on service/geopmdpy_test/TestActiveSessions.py:
```diff
@@ -0,0 +1,648 @@
+                mock.call(f'Warning: <geopm-service> {sess_path} has wrong user owner, it will be renamed to {renamed_path}'),
+                mock.call(f'Warning: <geopm-service> the wrong user owner was {bad_user.st_uid}')
+            ]
+            for cc, mc in zip(calls, mock_err.call_args_list):
+                self.assertEqual(cc, mc)
```
```suggestion
            mock_err.assert_has_calls(calls)
```

bgeltz commented on service/geopmdpy_test/TestActiveSessions.py:
```diff
@@ -0,0 +1,648 @@
+                mock.call(f'Warning: <geopm-service> {sess_path} has wrong group owner, it will be renamed to {renamed_path}'),
+                mock.call(f'Warning: <geopm-service> the wrong group owner was {bad_user.st_gid}')
+            ]
+            for cc, mc in zip(calls, mock_err.call_args_list):
+                self.assertEqual(cc, mc)
```
```suggestion
            mock_err.assert_has_calls(calls)
```

bgeltz commented on service/geopmdpy_test/TestActiveSessions.py:
```diff
@@ -0,0 +1,648 @@
+                mock.call(f'Warning: <geopm-service> session file was discovered with invalid permissions, will be ignored and removed: {os.path.join(sess_path, "session-2.json")}'),
+                mock.call(f'Warning: <geopm-service> the wrong permissions were {oct(0o644)}')
+            ]
+            for cc, mc in zip(calls, mock_err.call_args_list):
+                self.assertEqual(cc, mc)
```
```suggestion
            mock_err.assert_has_calls(calls)
```

bgeltz commented on service/geopmdpy_test/TestActiveSessions.py:
```diff
@@ -0,0 +1,648 @@
+                mock.call(f'Warning: <geopm-service> session file was discovered with invalid permissions, will be ignored and removed: {full_file_path}'),
+                mock.call(f'Warning: <geopm-service> the wrong user owner was {session_3_mock.st_uid}')
+            ]
+            for cc, mc in zip(calls, mock_err.call_args_list):
+                self.assertEqual(cc, mc)
```
```suggestion
            mock_err.assert_has_calls(calls)
```

bgeltz commented on service/geopmdpy_test/TestActiveSessions.py:
```diff
@@ -0,0 +1,648 @@
+                mock.call(f'Warning: <geopm-service> session file was discovered with invalid permissions, will be ignored and removed: {full_file_path}'),
+                mock.call(f'Warning: <geopm-service> the wrong group owner was {session_4_mock.st_gid}')
+            ]
+            for cc, mc in zip(calls, mock_err.call_args_list):
+                self.assertEqual(cc, mc)
```
```suggestion
            mock_err.assert_has_calls(calls)
```

bgeltz commented on service/geopmdpy_test/TestActiveSessions.py:
```diff
@@ -0,0 +1,648 @@
+                    contents["watch_id"]
+                )
+            else:
+                for cc, mc in zip(calls, mock_err.call_args_list):
+                    self.assertEqual(cc, mc)
```
```suggestion
                mock_err.assert_has_calls(calls)
```

bgeltz commented on service/geopmdpy_test/TestActiveSessions.py:
```diff
@@ -0,0 +1,648 @@
+        self.assertFalse(session.is_write_client(client_pid))
+
+    def create_json_file(self, directory, filename, contents, permissions=0o600):
+        """Create a json file that  matches the ActiveSessions._session_schema
```
This helper does create a JSON file, but it does not validate it against the session schema, so there are no guarantees that it matches the schema right now.

Was that intended?

bgeltz commented on service/geopmdpy_test/TestActiveSessions.py:
```diff
@@ -0,0 +1,648 @@
+                mock.call(f'Warning: <geopm-service> {sess_path} is a symbolic link, the link will be renamed to {renamed_path}'),
+                mock.call(f'Warning: <geopm-service> the symbolic link points to /root')
+            ]
+            for cc, mc in zip(calls, mock_err.call_args_list):
+                self.assertEqual(cc, mc)
```
```suggestion
            mock_err.assert_has_calls(calls)
```

bgeltz commented on service/geopmdpy_test/TestActiveSessions.py:
```diff
@@ -0,0 +1,648 @@
+             mock.patch('sys.stderr.write', return_value=None) as mock_err:
+            act_sess = ActiveSessions(sess_path)
+            renamed_path = f'{sess_path}-uuid4-INVALID'
+            calls = [
+                mock.call(f'Warning: <geopm-service> {sess_path} is not a directory, it will be renamed to {renamed_path}')
+            ]
+            for cc, mc in zip(calls, mock_err.call_args_list):
+                self.assertEqual(cc, mc)
```
```suggestion
            msg = f'Warning: <geopm-service> {sess_path} is not a directory, it will be renamed to {renamed_path}'
            mock_err.assert_called_with(msg)
```

bgeltz commented on service/geopmdpy_test/TestPlatformService.py:
```diff
@@ -366,58 +377,60 @@ def test_unlock_control(self):
+                self._mock_active_sessions.add_client.assert_called_with(client_pid, signals, controls, watch_id)
+            else:
+                self._mock_active_sessions.add_client.assert_not_called()
+
```
```suggestion
```

bgeltz commented on service/geopmdpy_test/TestPlatformService.py:
```diff
@@ -526,6 +545,7 @@ def test_start_batch(self):
         control_config = [(0, 0, con) for con in valid_controls]
 
         expected_result = (1234, "1234")
+
```
```suggestion
```

bgeltz commented on service/geopmdpy_test/TestActiveSessions.py:
```diff
@@ -217,11 +216,8 @@ def test_creation_file_not_dir(self):
-            for cc, mc in zip(calls, mock_err.call_args_list):
-                self.assertEqual(cc, mc)
+            msg = f'Warning: <geopm-service> {sess_path} is not a directory, it will be renamed to {renamed_path}'
+            mock_err.assert_called_with(msg)
```
assert_called_once_with
